data/abi/abi_encode_test.go

```diff
@@ -0,0 +1,1016 @@
+	})
+
+	// decoding test for tuple with static bool array
+	// expected type: (bool[2], bool[2])
+	// expected value: ({T, T}, {T, T})
+	/*
+	   input bytes:
+	   0b11000000      (first static bool array)
+	   0b11000000      (second static bool array)
+	*/
+	t.Run("static bool array tuple decoding", func(t *testing.T) {
+		boolArr := []bool{true, true}
+		boolValArr := make([]Value, 2)
+		for i := 0; i < 2; i++ {
+			boolValArr[i] = MakeBool(boolArr[i])
+		}
+		boolArrVal, err := MakeStaticArray(boolValArr)
+		require.NoError(t, err, "make bool static array should not return error")
+		tupleVal, err := MakeTuple([]Value{boolArrVal, boolArrVal})
+		require.NoError(t, err, "make tuple value should not return error")
+		encodedInput := []byte{
+			0b11000000,
+			0b11000000,
+		}
+		decoded, err := Decode(encodedInput, Type{
+			enumIndex:    Tuple,
+			staticLength: 2,
+			childTypes: []Type{
+				{
+					enumIndex:    ArrayStatic,
+					staticLength: 2,
+					childTypes:   []Type{MakeBoolType()},
+				},
+				{
+					enumIndex:    ArrayStatic,
+					staticLength: 2,
+					childTypes:   []Type{MakeBoolType()},
+				},
+			},
+		})
+		require.NoError(t, err, "decode tuple value should not return error")
+		require.Equal(t, tupleVal, decoded, "decoded tuple value do not match with expected")
+	})
+
+	// decoding test for tuple with static and dynamic bool array
+	// expected type: (bool[2], bool[])
+	// expected value: ({T, T}, {T, T})
+	/*
+	   input bytes:
+	   0b11000000      (first static bool array)
+	   0x00, 0x03      (second dynamic bool array starts at 3rd byte)
+	   0x00, 0x02      (dynamic bool array length 2)
+	   0b11000000      (second static bool array)
+	*/
+	t.Run("static/dynamic bool array tuple decoding", func(t *testing.T) {
+		boolArr := []bool{true, true}
+		boolValArr := make([]Value, 2)
+		for i := 0; i < 2; i++ {
+			boolValArr[i] = MakeBool(boolArr[i])
+		}
+		boolArrStaticVal, err := MakeStaticArray(boolValArr)
+		require.NoError(t, err, "make static bool array should not return error")
+		boolArrDynamicVal, err := MakeDynamicArray(boolValArr, MakeBoolType())
+		require.NoError(t, err, "make dynamic bool array should not return error")
+		tupleVal, err := MakeTuple([]Value{boolArrStaticVal, boolArrDynamicVal})
+		require.NoError(t, err, "make tuple for static/dynamic bool array should not return error")
+		encodedInput := []byte{
+			0b11000000,
+			0x00, 0x03,
+			0x00, 0x02, 0b11000000,
+		}
+		decoded, err := Decode(encodedInput, Type{
+			enumIndex:    Tuple,
+			staticLength: 2,
+			childTypes: []Type{
+				{
+					enumIndex:    ArrayStatic,
+					staticLength: 2,
+					childTypes:   []Type{MakeBoolType()},
+				},
+				{
+					enumIndex:  ArrayDynamic,
+					childTypes: []Type{MakeBoolType()},
+				},
+			},
+		})
+		require.NoError(t, err, "decode tuple for static/dynamic bool array should not return error")
+		require.Equal(t, tupleVal, decoded, "decoded tuple value do not match with expected")
+	})
```

Comment:
Given the current ABI implementation, it seems hard to make *concise* test case (i.e., define a test in a few lines), since it is tedious to make an ABI value, or an ABI type (because of the `err` return).

I added more comments, trying to explain away the attempt and mechanism behind the test.